contributors/design-proposals/kubeadm-upgrades.md

```diff
@@ -0,0 +1,350 @@
+
+**Decision**: Keep the logic inside of the kubeadm CLI (option 1) for the implementation in v1.8.0.
+
+We *might* revisit this and build an upgrade operator in future, either before GA of this functionality or after, in v2.
+
+**Rewrite manifests completely from config object or mutate existing manifest:**
+
+Instead of generating new manifests from a versioned configuration object, we could try to add "filters" to the existing manifests and apply different filters depending on what the upgrade looks like. This approach, to modify existing manifests in various ways depending on the version bump has some pros (modularity, simple mutating functions), but the matrix of functions and different policies would grow just too big for this kind of system, so we voted against this alternative in favor for the solution above.
+
+
+### Static Pod-hosted or Self-hosted control plane?
+
+A control plane can be set up in a couple of ways:
+
+ - Running the control plane components standalone; as systemd services for instance.
+   - kubeadm **does not** support handling any such cluster
+ - Running the control plane components in containers with kubelet as the babysitter; using Static Pods
+   - [Static Pods](#TODO) is a way to run standalone Pods on a node by dropping Pod manifests in a specific path.
+   - In kubeadm v1.4 to v1.8, this is the default way of setting up the control plane
+ - Running the control plane in Kubernetes-hosted containers as DaemonSets; aka. [Self-Hosting](#TODO)
+   - When creating a self-hosted cluster; kubeadm first creates a Static Pod-hosted cluster and then pivots to the self-hosted control plane.
+   - This is the default way to deploy the control plane since v1.9; but the user can opt-out if it and stick with the Static Pod-hosted cluster. 
+
+#### Static Pod-hosted control plane
+
+When the upgrade functionality is introduced in v1.8; the default way to set up the control plane is by using Static Pods.
+
+When the control plane is hosted using Static Pods, the actual upgrade flow looks like this:
+ - Gets the Mirror Pod manifests from the API Server (`kubectl -n kube-system get pod ${component}-${nodename} -ojson`), and sha256-hashes the JSON content.
+ - Writes new Static Pod manifests to a **temporary directory** (with the prefix `/etc/kubernetes/tmp/kubeadm-upgraded-manifests*`) based on the new configuration object built internally earlier.
+ - Creates a **backup directory** with the prefix `/etc/kubernetes/tmp/kubeadm-backup-manifests*`.
+   - Q: Why `/etc/kubernetes/tmp`?
+   - A: Possibly not very likely, but we concluded that there may be an attack area for computers where `/tmp` is shared and writable by all users.
+     We wouldn't want anyone to mock with the new Static Pod manifests being applied to the clusters. Hence we chose `/etc/kubernetes/tmp`, which is
+     root-only owned.
+ - In a loop for all control plane components:
```

Comment:
does the order of the control plane components matter?

Reply:
added a comment that order _shouldn't_ matter, but we do apiserver, ctlr-mgr, sched